### ISSUE TITLE
Doesn't interact correctly with `typing.overload`.

### ISSUE THREAD

**Issue by hoodmane** [tox-dev/sphinx-autodoc-typehints#296]
If you provide a `@typing.overload` for a function, the output is not correct. As far as I can tell, this problem is not fixable without changes to autodoc.

The is that in the presence of a `typing.overload`, autodoc will never emit an `'autodoc-process-signature'` event. You can see the logic here in `FunctionDocumenter.format_signature`:
https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/autodoc/__init__.py?plain=1#L2193
If `self.analyzer.overloads` has overloads for this function, it uses overloaded signatures, otherwise it calls `super().format_signature`. In this case `super().format_signature` is `Documenter.format_signature` which actually emits the `'autodoc-process-signature' event:
https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/autodoc/__init__.py?plain=1#L513\

So when overloads exist, we fail to override default behavior. An easy fix for this is:
```py
from sphinx.ext.autodoc import FunctionDocumenter
del FunctionDocumenter.format_signature
```

<details>
<summary>Example module</summary>

```py
from typing import overload


@overload
def f(a : int, b : int) -> None:
    ...

@overload
def f(a : str, b : str) -> None:
    ...

def f(a : int | str, b : int | str) -> None:
    """
    f does the thing. The arguments can either be ints or strings but they must
    both have the same type.

    Parameters
    ----------
    a:
        The first thing
    b:
        The second thing
    """
```

</details>

#### Actual output
<details>
<summary>Actual output</summary>

![Screenshot from 2023-01-16 19-29-54](https://user-images.githubusercontent.com/8739626/212803861-3024571c-2db6-49f9-819e-3da899bd8a16.png)

</details>

#### Desired output 

achieved with `del FunctionDocumenter.format_signature`

<details>
<summary>Desired output</summary>

![Screenshot from 2023-01-16 19-30-47](https://user-images.githubusercontent.com/8739626/212803874-5b97a15b-2602-46a2-bd45-f55eb218700f.png)

</details>

**Comment by hoodmane** [comment]
I guess you also have to `del MethodDocumenter.format_signature`.

**Comment by gaborbernat** [comment]
PR trying to address this is welcome 🤗

**Comment by hoodmane** [comment]
Are you okay with adding:
```py
from sphinx.ext.autodoc import FunctionDocumenter, MethodDocumenter

del FunctionDocumenter.format_signature
del MethodDocumenter.format_signature
```
to `setup`?

**Comment by gaborbernat** [comment]
I'm willing to try.

**Comment by hoodmane** [comment]
Okay, PR incoming.

**Comment by hoodmane** [comment]
Closed by #297.
